In [2]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sns
from sklearn import datasets
from sklearn.linear_model import LinearRegression
from sklearn.linear_model import LogisticRegression
from sklearn import tree
from sklearn.svm import SVC
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import train_test_split
from sklearn.model_selection import KFold
from sklearn.model_selection import StratifiedKFold
from sklearn.model_selection import cross_val_score
from sklearn.preprocessing import LabelEncoder
from sklearn.metrics import confusion_matrix
from sklearn import tree
import math

In [2]:
digits = datasets.load_digits()
dir(digits)

['DESCR', 'data', 'feature_names', 'frame', 'images', 'target', 'target_names']

In [3]:
X_train, X_test, y_train, y_test = train_test_split(digits.data, digits.target, test_size=0.3, random_state=10)

In [6]:
lr = LogisticRegression(max_iter=10000)
lr.fit(X_train, y_train)
lr.score(X_test, y_test)

0.9648148148148148

In [18]:
svm = SVC()
svm.fit(X_train, y_train)
svm.score(X_test, y_test)

0.9833333333333333

In [17]:
rf = RandomForestClassifier(n_estimators=100)
rf.fit(X_train, y_train)
rf.score(X_test, y_test)

0.9722222222222222

In [20]:
kf = KFold(n_splits=3)
kf

KFold(n_splits=3, random_state=None, shuffle=False)

In [21]:
for train_index, test_index in kf.split([1, 2, 3, 4, 5, 6, 7, 8, 9]):
    print(train_index, test_index)

[3 4 5 6 7 8] [0 1 2]
[0 1 2 6 7 8] [3 4 5]
[0 1 2 3 4 5] [6 7 8]


In [22]:
def get_score(model, X_train, X_test, y_train, y_test):
    model.fit(X_train, y_train)
    return model.score(X_test, y_test)

In [25]:
get_score(LogisticRegression(max_iter=10000), X_train, X_test, y_train, y_test)

0.9648148148148148

In [26]:
get_score(SVC(), X_train, X_test, y_train, y_test)

0.9833333333333333

In [27]:
get_score(RandomForestClassifier(), X_train, X_test, y_train, y_test)

0.975925925925926

In [29]:
folds = StratifiedKFold(n_splits=3)

In [35]:
scores_lr = []
scores_svm = []
scores_rf = []

for train_index, test_index in folds.split(digits.data, digits.target):
    X_train, X_test, y_train, y_test = digits.data[train_index], digits.data[test_index], \
                                       digits.target[train_index], digits.target[test_index]
    scores_lr.append(get_score(LogisticRegression(max_iter=10000), X_train, X_test, y_train, y_test))
    scores_svm.append(get_score(SVC(), X_train, X_test, y_train, y_test))
    scores_rf.append(get_score(RandomForestClassifier(n_estimators=40), X_train, X_test, y_train, y_test))
print(f"Logistic Regression scores: {scores_lr}")
print(f"SVM scores: {scores_svm}")
print(f"Random Forest scores: {scores_rf}")

Logistic Regression scores: [0.9248747913188647, 0.9398998330550918, 0.9232053422370617]
SVM scores: [0.9649415692821369, 0.9799666110183639, 0.9649415692821369]
Random Forest scores: [0.9315525876460768, 0.9532554257095158, 0.9298831385642737]


In [38]:
cross_val_score(LogisticRegression(max_iter=10000), digits.data, digits.target)

array([0.925     , 0.875     , 0.93871866, 0.93314763, 0.89693593])

In [39]:
cross_val_score(SVC(), digits.data, digits.target)

array([0.96111111, 0.94444444, 0.98328691, 0.98885794, 0.93871866])

In [43]:
cross_val_score(RandomForestClassifier(n_estimators=40), digits.data, digits.target)

array([0.93611111, 0.9       , 0.9637883 , 0.94428969, 0.92479109])

In [42]:
cross_val_score(RandomForestClassifier(n_estimators=5), digits.data, digits.target)

array([0.88055556, 0.81388889, 0.86629526, 0.90529248, 0.84958217])

In [44]:
cross_val_score(RandomForestClassifier(n_estimators=15), digits.data, digits.target)

array([0.925     , 0.88611111, 0.94707521, 0.94428969, 0.8913649 ])

In [45]:
cross_val_score(RandomForestClassifier(n_estimators=60), digits.data, digits.target)

array([0.93333333, 0.90555556, 0.95264624, 0.96100279, 0.91643454])

In [46]:
scores = cross_val_score(RandomForestClassifier(n_estimators=100), digits.data, digits.target)
np.average(scores)

0.9388037759207677

### Homework

Logistic Regression  
SVM  
Decision Tree  
Random Forest

In [3]:
iris = datasets.load_iris()
data = pd.DataFrame(iris.data, columns=iris.feature_names)
data.head()

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
0                5.1               3.5                1.4               0.2
1                4.9               3.0                1.4               0.2
2                4.7               3.2                1.3               0.2
3                4.6               3.1                1.5               0.2
4                5.0               3.6                1.4               0.2

In [4]:
data['target'] = iris.target
data.head()

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)  \
0                5.1               3.5                1.4               0.2   
1                4.9               3.0                1.4               0.2   
2                4.7               3.2                1.3               0.2   
3                4.6               3.1                1.5               0.2   
4                5.0               3.6                1.4               0.2   

   target  
0       0  
1       0  
2       0  
3       0  
4       0

In [12]:
X = data.drop(['target'], axis=1)
y = data['target']
print(X.head())
print('*'*75)
print(y.head())

   sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
0                5.1               3.5                1.4               0.2
1                4.9               3.0                1.4               0.2
2                4.7               3.2                1.3               0.2
3                4.6               3.1                1.5               0.2
4                5.0               3.6                1.4               0.2
***************************************************************************
0    0
1    0
2    0
3    0
4    0
Name: target, dtype: int32


In [14]:
lr_scores = cross_val_score(LogisticRegression(max_iter=10000), X, y)
avg_lr_score = np.average(lr_scores)
avg_lr_score

0.9733333333333334

In [15]:
dt_scores = cross_val_score(tree.DecisionTreeClassifier(), X, y)
avg_dt_score = np.average(dt_scores)
avg_dt_score

0.9600000000000002

In [16]:
svm_scores = cross_val_score(SVC(), X, y)
avg_svm_score = np.average(svm_scores)
avg_svm_score

0.9666666666666666

In [17]:
rf_scores = cross_val_score(RandomForestClassifier(), X, y)
avg_rf_score = np.average(rf_scores)
avg_rf_score

0.9666666666666668

The best score is from Logistic regression model: 0.973